### PR TITLE
Create backup of databases before opening contexts

### LINKED ISSUE #16531: Debug build of lazer crashes on launch when migrating to Realm
# Steps to repro

1. Delete lazer data folder.
2. Checkout 2ad0ea35be5c170546b8cfca8f7cae0939567bcd and run that version. (Probably not related to that specific commit, it's just the commit I happened to be on when I first found the bug.)
3. Choose to start fresh at the specified location when prompted.
4. Close the game.
5. Check out master (at 1a207251628d50ae701ec73a4c47f0f27ca29930 right now) and run the game.
6. Game fails to launch with an IOException:
<details>
<summary>Stack Trace</summary>

```
2022-01-20 13:43:20 [verbose]: Beginning realm file store cleanup
2022-01-20 13:43:20 [verbose]: Finished realm file store cleanup (0 of 0 deleted)
2022-01-20 13:43:21 [important]: Your development database has been fully migrated to realm. If you switch back to a pre-realm branch and need your previous database, rename the backup file back to "client.db".
2022-01-20 13:43:21 [important]: 
2022-01-20 13:43:21 [important]: Note that doing this can potentially leave your file store in a bad state.
2022-01-20 13:43:21 [error]: An unhandled error has occurred.
2022-01-20 13:43:21 [error]: System.IO.IOException: The process cannot access the file 'D:\Games\osu!lazer\data\client.db' because it is being used by another process.
2022-01-20 13:43:21 [error]: at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
2022-01-20 13:43:21 [error]: at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
2022-01-20 13:43:21 [error]: at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
2022-01-20 13:43:21 [error]: at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
2022-01-20 13:43:21 [error]: at System.IO.File.Open(String path, FileMode mode, FileAccess access, FileShare share)
2022-01-20 13:43:21 [error]: at osu.Framework.Platform.NativeStorage.GetStream(String path, FileAccess access, FileMode mode)
2022-01-20 13:43:21 [error]: at osu.Game.IO.WrappedStorage.GetStream(String path, FileAccess access, FileMode mode) in D:\Projects\OSU\osu\osu.Game\IO\WrappedStorage.cs:line 71
2022-01-20 13:43:21 [error]: at osu.Game.Database.DatabaseContextFactory.CreateBackup(String backupFilename) in D:\Projects\OSU\osu\osu.Game\Database\DatabaseContextFactory.cs:line 157
2022-01-20 13:43:21 [error]: at osu.Game.Database.EFToRealmMigrator.ensureBackup() in D:\Projects\OSU\osu\osu.Game\Database\EFToRealmMigrator.cs:line 409
2022-01-20 13:43:21 [error]: at osu.Game.Database.EFToRealmMigrator.migrateBeatmaps(OsuDbContext ef) in D:\Projects\OSU\osu\osu.Game\Database\EFToRealmMigrator.cs:line 80
2022-01-20 13:43:21 [error]: at osu.Game.Database.EFToRealmMigrator.Run() in D:\Projects\OSU\osu\osu.Game\Database\EFToRealmMigrator.cs:line 45
2022-01-20 13:43:21 [error]: at osu.Game.OsuGameBase.load(ReadableKeyCombinationProvider keyCombinationProvider) in D:\Projects\OSU\osu\osu.Game\OsuGameBase.cs:line 198
```
</details>

Here's how the data folder looks like after the crash: (Backup files are missing)
image

I can confirm that the migration itself is successful since the game launches just fine if I delete `client.db`. Modified settings are still persisted. It is only crashing in the backup creation part.

# Logs
[performance.log](https://github.com/ppy/osu/files/7905701/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7905702/runtime.log)
[database.log](https://github.com/ppy/osu/files/7905703/database.log)
[network.log](https://github.com/ppy/osu/files/7905704/network.log)

### PR DESCRIPTION
**Pull request by peppy**
Attempt to avoid file IO issues.

Closes #16531.

### CODE REVIEW

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 1]
```diff
@@ -303,15 +303,12 @@ public WorkingBeatmap GetWorkingBeatmap(BeatmapInfo? importedBeatmap)
         {
             // Detached sets don't come with files.
             // If we seem to be missing files, now is a good time to re-fetch.
-            if (importedBeatmap?.BeatmapSet?.Files.Count == 0)
+            if (importedBeatmap?.IsManaged == false && importedBeatmap.BeatmapSet?.Files.Count == 0)
             {
-                using (var realm = contextFactory.CreateContext())
-                {
-                    var refetch = realm.Find<BeatmapInfo>(importedBeatmap.ID)?.Detach();
+                var refetch = contextFactory.Context.Find<BeatmapInfo>(importedBeatmap.ID)?.Detach();
```
Comment (bdach) [inline osu.Game/Beatmaps/BeatmapManager.cs:308]: `GetWorkingBeatmap()` is routinely called outside of the update thread, so this change can't work in general (as evidenced by CI).

**Review by bdach — DISMISSED** [review]

File: osu.Game/Beatmaps/BeatmapManager.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Let's pretend that didn't happen.

**Comment by bdach** [comment]
Code seems fine. In my testing I can see that this does appear to fix #16531 on my machine, but I cannot reproduce [the second trace](https://github.com/ppy/osu/issues/16531#issuecomment-1017619649) @Henry-YSLin posted, and because this branch has been force-pushed since then, I cannot easily check whether I would get the same exception on the commits pre-force-push.

So essentially the first crash looks fixed, but when it comes to the second crash, I cannot either confirm or deny whether it is fixed with the addition of the retry logic.

**Comment by peppy** [comment]
> Code seems fine. In my testing I can see that this does appear to fix #16531 on my machine, but I cannot reproduce [the second trace](https://github.com/ppy/osu/issues/16531#issuecomment-1017619649) @Henry-YSLin posted, and because this branch has been force-pushed since then, I cannot easily check whether I would get the same exception on the commits pre-force-push.
> 
> 
> 
> So essentially the first crash looks fixed, but when it comes to the second crash, I cannot either confirm or deny whether it is fixed with the addition of the retry logic.

The force push removed unrelated code. The two commits they remain are the two requested attempts (second commit was after the first one reported to not help).

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**